Client/Frontend/ContentBlocker/ContentBlockerHelper.swift

```diff
@@ -0,0 +1,303 @@
+                    }
+                }
+            }
+        }
+
+        switch readEnabledPref() {
+        case .off:
+            if reloadTab {
+                self.tab?.reload()
+            }
+            return
+        case .on:
+            addRules()
+        case .onInPrivateBrowsing:
+            if tab?.isPrivate ?? false {
+                addRules()
+            } else {
+                self.tab?.reload()
+            }
+        }
+    }
+
+    func removeAllFromTab() {
+        tab?.webView?.configuration.userContentController.removeAllContentRuleLists()
+    }
+
+    fileprivate func addToTab(contentRuleList: WKContentRuleList?, error: Error?) {
+        if let rules = contentRuleList {
+            tab?.webView?.configuration.userContentController.add(rules)
+        } else {
+            print("Content blocker load error: " + (error?.localizedDescription ?? "empty rules"))
+            assert(false)
+        }
+    }
+}
+
+// MARK: Private initialization code
+// The rule store can compile JSON rule files into a private format which is cached on disk.
+// On app boot, we need to check if the ruleStore's data is out-of-date, or if the names of the rule files
+// no longer match. Finally, any JSON rule files that aren't in the ruleStore need to be compiled and stored in the
+// ruleStore.
+@available(iOS 11, *)
+extension ContentBlockerHelper {
+    fileprivate func loadJsonFromBundle(forResource file: String, completion: @escaping (_ jsonString: String) -> Void) {
+        DispatchQueue.global().async {
+            guard let path = Bundle.main.path(forResource: file, ofType: "json"),
+                let source = try? NSString(contentsOfFile: path, encoding: String.Encoding.utf8.rawValue) as String else {
+                    return
+            }
+            DispatchQueue.main.async {
+                completion(source)
+            }
+        }
+    }
+
+    fileprivate func lastModifiedSince1970(forFileAtPath path: String) -> Timestamp? {
+        do {
+            let url = URL(fileURLWithPath: path)
+            let attr = try FileManager.default.attributesOfItem(atPath: url.path)
+            guard let date = attr[FileAttributeKey.modificationDate] as? Date else { return nil }
+            return UInt64(1000.0 * date.timeIntervalSince1970)
+        } catch {
+            return nil
+        }
+    }
+
+    fileprivate func dateOfMostRecentBlockerFile() -> Timestamp {
+        let blocklists = blocklistBasic + blocklistStrict
+        return blocklists.reduce(Timestamp(0)) { result, filename in
+            guard let path = Bundle.main.path(forResource: filename, ofType: "json") else { return result }
+            let date = lastModifiedSince1970(forFileAtPath: path) ?? 0
+            return date > result ? date : result
+        }
+    }
+
+    fileprivate func removeAllRulesInStore(completion: @escaping () -> Void) {
+        guard let ruleStore = ruleStore else { return }
```

Comment:
Intended, as I am not sure exactly what to do with the fact that Apple has made the init of the rule store failable, and there is no doc as to in what case this would be nil. In the case the ruleStore is nil, this code is designed to no-op and avoid crashing. Thoughts?